File: src/main/java/sage/domain/service/BlogPostService.java
package sage.domain.service;

import java.util.Collection;
import java.util.Date;

import httl.util.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import sage.domain.commons.BadArgumentException;
import sage.domain.commons.DomainRuntimeException;
import sage.domain.commons.IdCommons;
import sage.domain.repository.BlogRepository;
import sage.domain.repository.TagRepository;
import sage.domain.repository.UserRepository;
import sage.domain.search.SearchBase;
import sage.entity.Blog;
import sage.transfer.BlogView;

@Service
@Transactional
public class BlogPostService {
  private static final int BLOG_TITLE_MAX_LEN = 100, BLOG_CONTENT_MAX_LEN = 10000;
  private static final BadArgumentException BAD_INPUT_LENGTH = new BadArgumentException(
      "输入长度不正确(标题1~100字,内容1~10000字)");

  @Autowired
  private SearchBase searchBase;
  @Autowired
  private BlogRepository blogRepo;
  @Autowired
  private UserRepository userRepo;
  @Autowired
  private TagRepository tagRepo;

  public Blog post(long userId, String title, String content, Collection<Long> tagIds) {
    checkLength(title, content);
    Blog blog = new Blog(title, content, userRepo.load(userId), new Date(), tagRepo.byIds(tagIds));
    escapeAndSetText(blog);

    blogRepo.save(blog);
    searchBase.index(blog.getId(), new BlogView(blog));
    return blog;
  }

  public Blog edit(long userId, long blogId, String title, String content, Collection<Long> tagIds) {
    checkLength(title, content);
    Blog blog = blogRepo.get(blogId);
    if (!IdCommons.equal(blog.getAuthor().getId(), userId)) {
      throw new DomainRuntimeException("User[%d] is not the author of Blog[%d]", userId, blogId);
    }
    blog.setTitle(title);
    blog.setContent(content);
    blog.setModifiedTime(new Date());
    blog.setTags(tagRepo.byIds(tagIds));
    escapeAndSetText(blog);

    blogRepo.update(blog);
    searchBase.index(blog.getId(), new BlogView(blog));
    return blog;
  }

  public void delete(long userId, long blogId) {
    Blog blog = blogRepo.get(blogId);
    if (!IdCommons.equal(blog.getAuthor().getId(), userId)) {
      throw new DomainRuntimeException("User[%d] is not the author of Blog[%d]", userId, blogId);
    }
    blogRepo.delete(blog);
    searchBase.delete(BlogView.class, blog.getId());
  }

  private void checkLength(String title, String content) {
    if (title.isEmpty() || title.length() > BLOG_TITLE_MAX_LEN
        || content.isEmpty() || content.length() > BLOG_CONTENT_MAX_LEN) {
      throw BAD_INPUT_LENGTH;
    }
  }

  private void escapeAndSetText(Blog blog) {
    String title = StringUtils.escapeXml(blog.getTitle());
    String content = StringUtils.escapeXml(blog.getContent()).replace("\n", "  \n");
    blog.setTitle(title);
    blog.setContent(content);
  }
}

File: src/main/kotlin/sage/service/BlogPostService.kt
package sage.service

import httl.util.StringUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import sage.domain.commons.BadArgumentException
import sage.domain.commons.DomainException
import sage.domain.search.SearchBase
import sage.entity.Blog
import sage.entity.Tag
import sage.entity.User
import sage.transfer.BlogView

@Service
class BlogPostService
@Autowired constructor(private val searchBase: SearchBase) {

  fun post(userId: Long, title: String, content: String, tagIds: Collection<Long>): Blog {
    checkLength(title, content)
    val blog = Blog(title, content, User.ref(userId), Tag.multiGet(tagIds))
    blog.escapeAndSetText().save()
    searchBase.index(blog.id, BlogView(blog))
    return blog
  }

  fun edit(userId: Long, blogId: Long, title: String, content: String, tagIds: Collection<Long>): Blog {
    checkLength(title, content)
    val blog = get(blogId)
    if (userId != blog.author.id) {
      throw DomainException("User[%s] is not the author of Blog[%s]", userId, blogId)
    }
    blog.title = title
    blog.content = content
    blog.tags = Tag.multiGet(tagIds)
    blog.escapeAndSetText().update()
    searchBase.index(blog.id, BlogView(blog))
    return blog
  }

  fun delete(userId: Long, blogId: Long) {
    val blog = get(blogId)
    if (userId != blog.author.id) {
      throw DomainException("User[%d] is not the author of Blog[%d]", userId, blogId)
    }
    blog.delete()
    searchBase.delete(BlogView::class.java, blog.id)
  }

  private fun get(blogId: Long) =
      Blog.byId(blogId) ?: throw DomainException("Blog{%s] does not exist", blogId)

  private fun checkLength(title: String, content: String) {
    if (title.isEmpty() || title.length > BLOG_TITLE_MAX_LEN
        || content.isEmpty() || content.length > BLOG_CONTENT_MAX_LEN) {
      throw BAD_INPUT_LENGTH
    }
  }

  private fun Blog.escapeAndSetText(): Blog {
    title = StringUtils.escapeXml(title)
    content = StringUtils.escapeXml(content).replace("\n", "  \n")
    return this
  }

  companion object {
    private val BLOG_TITLE_MAX_LEN = 100
    private val BLOG_CONTENT_MAX_LEN = 10000
    private val BAD_INPUT_LENGTH = BadArgumentException(
        "输入长度不正确(标题1~100字,内容1~10000字)")
  }
}
